File: python/pixel_perfect_circles/length_utils.py
from typing import List, Tuple


def get_has_intersection_actual_size(arc_size: int, lengths: List[int]) -> Tuple[bool, int]:
    has_intersection = sum(lengths) == arc_size - len(lengths) + 1
    intersection_cells = 1 if has_intersection else 0
    actual_size = sum(lengths) + len(lengths) - intersection_cells
    return has_intersection, actual_size


def is_possible_lengths(arc_size: int, lengths: List[int]) -> int:
    _, actual_size = get_has_intersection_actual_size(arc_size, lengths)
    return actual_size == arc_size


def get_circle_by_lengths(size: int, lengths: List[int]) -> List[List[bool]]:
    arc_size = (size + 1) // 2
    has_intersection, actual_size = get_has_intersection_actual_size(arc_size, lengths)
    if actual_size != arc_size:
        print(f"Could not get grid by lengths, {actual_size} != {arc_size}")
        return []

    grid = []
    sum_length = -(size % 2)
    for length in lengths:
        sum_length += length * 2
        empty_length = (size - sum_length) // 2
        line = [False] * empty_length + [True] * sum_length + [False] * empty_length
        grid.append(line)

    for i, length in enumerate(reversed(lengths)):
        if has_intersection and i == 0:
            length -= 1
            if length == 0:
                continue
        else:
            sum_length += 2

        empty_length = (size - sum_length) // 2
        line = [False] * empty_length + [True] * sum_length + [False] * empty_length
        grid.extend([line] * length)

    for i in range(size // 2 - 1, -1, -1):
        grid.append(grid[i])

    return grid


def is_monotonous_lengths(lengths: List[int]) -> bool:
    prev_length = lengths[0]
    for length in lengths:
        if length > prev_length:
            return False

        prev_length = length

    return True


def filter_monotonous_lengths(options: List[List[int]]) -> List[List[int]]:
    return list(filter(is_monotonous_lengths, options))


def get_lengths_options(arc_length: int) -> List[List[int]]:
    options = _get_lengths_options(arc_length)
    return [
        option for option in options
        if is_possible_lengths(arc_length, option)
    ]


def _get_lengths_options(remaining_length: int) -> List[List[int]]:
    if remaining_length <= 0:
        return [[]]

    if remaining_length == 1:
        return [[1]]

    result = []
    for length in range(1, remaining_length):
        result.extend([
            [length] + lengths
            for lengths in _get_lengths_options(remaining_length - length - 1)
        ])

    result.append([remaining_length])
    return result

File: js/length-utils.js
const lengthsCache = new Map();

function getOptimalLengthsSlow(size, lengthOptions) {
	let minAreaDeviation = +Infinity;
	let minLengths = lengthOptions[0];
	for (const lengths of lengthOptions) {
		let areaDeviation = getAreaDeviationByLengths(size, lengths);
		if (areaDeviation < minAreaDeviation) {
			minAreaDeviation = areaDeviation;
			minLengths = lengths;
		}
	}
	return minLengths;
}

function getOptimalLengths(size, startLengths) {
	let initLengths = startLengths.sort((a, b) => b - a);
	let initAreaDeviation = getAreaDeviationByLengths(size, initLengths);
	let minInfo = { minLengths: initLengths, minAreaDeviation: initAreaDeviation };
	let lengths = [];
	while (lengths.length !== minInfo.minLengths.length || lengths.some((val, i, arr) => val !== minInfo.minLengths[i])) {
		lengths = minInfo.minLengths;

		let addIndices = [];
		let subtractIndices = [];
		for (let i = 0; i < lengths.length; i++) {
			if (i == 0 || lengths[i - 1] > lengths[i]) {
				addIndices.push(i);
			}
			if ((i == lengths.length - 1 || lengths[i] > lengths[i + 1]) && lengths[i] != 1) {
				subtractIndices.push(i);
			}
		}

		for (let i of addIndices) {
			lengths[i]++;
			tryUpdateMinAreaDeviation(size, lengths, minInfo);

			if (lengths[lengths.length - 1] == 1) {
				lengths.pop();
				tryUpdateMinAreaDeviation(size, lengths, minInfo);
				lengths.push(1);
			}

			for (let j of subtractIndices) {
				if (i >= j)
					continue;

				lengths[j]--;
				if (lengths[j] == 0)
					lengths.pop();

				tryUpdateMinAreaDeviation(size, lengths, minInfo);

				if (j >= lengths.length)
					lengths.push(0);
				lengths[j]++;
			}
			lengths[i]--;
		}
		for (let i of subtractIndices) {
			lengths[i]--;
			tryUpdateMinAreaDeviation(size, lengths, minInfo);

			lengths.push(1);
			tryUpdateMinAreaDeviation(size, lengths, minInfo);
			lengths.pop();

			for (let j of addIndices) {
				if (i >= j)
					continue;

				lengths[j]++;
				if (isMonotonousLengths(lengths)) {
					tryUpdateMinAreaDeviation(size, lengths, minInfo);
				}
				lengths[j]--;
			}
			lengths[i]++;
		}
		console.log(size + ": " + minInfo.minLengths + " vs " + lengths);
	}
	return minInfo.minLengths;
}

function getAreaDeviationByLengths(size, lengths) {
	return getAreaDeviation(size, getCircleByLengths(size, lengths));
}

function getCircleByLengths(size, lengths) {
	let arcSize = Math.floor((size + 1) / 2);
	let hasIntersectionCell = hasIntersection(arcSize, lengths);
	let actualSize = getActualSize(arcSize, lengths);
	if (actualSize != arcSize) {
		console.log("Could not get grid by lengths, " + actualSize + " != " + arcSize);
		return [];
	}

	let grid = [];
	let sumLength = -(size % 2);
	for (const length of lengths) {
		sumLength += length * 2;
		let line = getGridLine(size, sumLength);
		grid.push(line);
	}

	for (let i = 0; i < lengths.length; i++) {
		let length = lengths[lengths.length - 1 - i];
		if (hasIntersectionCell && i == 0) {
			length -= 1;
			if (length == 0)
				continue;
		}
		else {
			sumLength += 2;
		}

		let line = getGridLine(size, sumLength);
		for (let j = 0; j < length; j++)
			grid.push(line);
	}

	for (let i = Math.floor(size / 2) - 1; i >= 0; i--)
		grid.push(grid[i]);

	return grid;
}

function getGridLine(size, sumLength) {
	let emptyLength = Math.floor((size - sumLength) / 2);
	let line = [];
	for (let i = 0; i < size; i++) {
		line.push(i >= emptyLength && i < size - emptyLength);
	}

	return line;
}

function getDefaultLengths(size) {
	if (size <= 2)
		return [1];

	let radius = size / 2;
	let i = 0;
	let j = Math.floor(size / 2 - 1);
	let res = [];
	while (true) {
		let y = i - radius + 0.5;
		let maxX = Math.sqrt(radius * radius - y * y);
		let max_j = Math.floor(maxX + radius - 0.5);
		// console.log(size, i, j, max_j, maxX, y);
		if (y >= 0.0 || size - max_j <= i)
			break;

		res.push(max_j - j);
		j = max_j;
		i += 1;
	}
	return res;
}

function getLengthOptions(size, monotonous) {
	let arcSize = Math.floor((size + 1) / 2);
	let options = _getLengthOptions(arcSize);
	if (!monotonous)
		return options;

	let res = [];
	for (const option of options)
		if (isMonotonousLengths(option))
			res.push(option);

	return res;
}


function _getLengthOptions(remainingLength) {
	if (remainingLength <= 0)
		return [[]];

	if (remainingLength == 1)
		return [[1]];

	if (lengthsCache.has(remainingLength)) {
		return lengthsCache.get(remainingLength);
	}

	let result = [[remainingLength]];
	for (let length = remainingLength - 1; length >= 1; length--) {
		for (const lengths of _getLengthOptions(remainingLength - length - 1)) {
			let newLengths = [length];
			newLengths.push(...lengths);
			result.push(newLengths);
		}
	}

	lengthsCache.set(remainingLength, result);
	return result;
}

function isPossibleLengths(arcSize, lengths) {
	let actualSize = getActualSize(arcSize, lengths);
	return actualSize == arcSize;
}

function isMonotonousLengths(lengths) {
	let prevLength = lengths[0];
	for (const length of lengths) {
		if (length > prevLength)
			return false;

		prevLength = length;
	}

	return true;
}

function hasIntersection(arcSize, lengths) {
	return sum(lengths) == arcSize - lengths.length + 1;
}

function getActualSize(arcSize, lengths) {
	let intersectionCells = hasIntersection(arcSize, lengths) ? 1 : 0;
	let actualSize = sum(lengths) + lengths.length - intersectionCells;
	return actualSize;
}

function sum(numberArray) {
	return numberArray.reduce((accumulator, currentValue) => {
		return accumulator + currentValue
	}, 0);
}

function tryUpdateMinAreaDeviation(size, lengths, minInfo) {
	let arcSize = Math.floor((size + 1) / 2);
	if (isPossibleLengths(arcSize, lengths)) {
		let areaDeviation = getAreaDeviationByLengths(size, lengths);
		if (areaDeviation < minInfo.minAreaDeviation) {
			minInfo.minAreaDeviation = areaDeviation;
			minInfo.minLengths = [...lengths];
		}
	}

}